src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -97,11 +97,17 @@ abstract class RefChecks extends Transform {
     private val checkedCombinations = mutable.Map[List[Symbol], (Symbol, Type)]()
     private def notYetCheckedOrAdd(rt: RefinedType, currentBase: Symbol) = {
       val seen = checkedCombinations.get(rt.parents.map(_.typeSymbol)).exists {
-        case (prevBase, prevTp) => currentBase.isSubClass(prevBase) && rt =:= prevTp.asSeenFrom(currentBase.thisType, prevBase)
+        case (prevBase, prevTp) =>
+          val isSub = (currentBase, prevBase) match {
+            case (cRef: RefinementClassSymbol, pRef: RefinementClassSymbol) =>
+              cRef.info.parents.map(_.typeSymbol) == pRef.info.parents.map(_.typeSymbol)
+            case _ =>
+              currentBase.isSubClass(prevBase)
+          }
+          val sameTp = rt =:= prevTp.asSeenFrom(currentBase.thisType, prevBase)
+          isSub && sameTp
```

Comment:
I'm wondering if instead we should just return "true" (== "not yet checked" for the `case (cRef: RefinementClassSymbol, pRef: RefinementClassSymbol)`.  If I understand this correctly, this is a caching optimization, so returning false would always be correct, at the risk of repeating some work.

Reply:
I think it's rather a visit log to stop recursion than a cache; that said

```scala
    private def notYetCheckedOrAdd(rt: RefinedType, currentBase: Symbol) = currentBase match {
      case _: RefinementClassSymbol => false
      case _ => // old code
    }
```

passes the test and doesn't break any existing ones. So maybe it's correct to assume that once the SymbolPair.base is a `RefinementClassSymbol` we don't need to keep calling `checkAllOverrides`? I'm not too sure. We're in

```
def checkOverride(pair: SymbolPair)
  ...
  if (member.isStable)
    ...
    lowType.normalize.resultType match {
      case rt: RefinedType if notYetCheckedOrAdd(rt, pair.base) =>
        ...
        checkAllOverrides(rt.typeSymbol) // needed if `pair.base` is `RefinementClassSymbol`?
```

The bug we're fixing is an endless recursion through this `checkAllOverrides` call.

Reply:
Let's go with the version you proposed... its hard to trust the test coverage enough to short circuit.